qiskit_dynamics/perturbation/multiset_utils.py

```diff
@@ -46,15 +46,55 @@ def _multiset_to_sorted_list(multiset: Multiset) -> List:
     return sorted_list
 
 
-def _sorted_multisets(multisets: List[Multiset]) -> List[Multiset]:
-    """Sort in non-decreasing order according to:
+class _MultisetSort:
+    """Dummy class for usage as a key when sorting Multiset instances. This assumes the entries
+    if the multisets can themselves be sorted.
+    """
+    __slots__ = "multiset",
+
+    def __init__(self, multiset: Multiset):
+        self.multiset = multiset
+    
+    def __lt__(self, other: Multiset) -> bool:
+        """Implements an ordering on multisets.
+        
+        This orders first according to length (the number of elements in each multiset). If self and
+        other are the same length, self < other if, when written as fully expanded and sorted lists,
+        self < other in lexicographic ordering. E.g. it holds that Multiset({0: 2, 1: 1}) <
+        Multiset({0: 1, 1: 2}), as the list versions are [0, 0, 1], and [0, 1, 1]. Here, the first
+        element of each is 0, so there is no comparison, but the second element of the first is <
+        the second, hence the ordering on the multisets.
+        """
+        if len(self.multiset) < len(other.multiset):
+            return True
+
+        if len(other.multiset) < len(self.multiset):
+            return False
+
+        unique_self = set(self.multiset.distinct_elements())
+        unique_other = set(other.multiset.distinct_elements())
+        unique_entries = list(unique_self.union(unique_other))
+        unique_entries.sort()
+
+        for element in unique_entries:
+            self_count = self.multiset[element]
+            other_count = other.multiset[element]
 
-    ms1 <= ms2 if len(ms1) < len(ms2), or if
-    len(ms1) == len(ms2) and if
-    str(_multiset_to_sorted_list(ms1)) <= str(_multiset_to_sorted_list(ms2)).
+            if self_count < other_count:
+                return False
+
+            if self_count > other_count:
+                return True
```

Comment:
> This is probably stylistic more than anything, but it does reduce the number of ifs (same can be applied to higher up in the function). BTW, without understanding the context at all, I'm slightly surprised to see True being returned if self_count > other_count. Intentional?

Yes this is intentional. This is the "lexicographic" ordering. E.g. for multisets `a = Multiset({0: 2, 1: 1})` and `b = Multiset({0: 1, 1: 2})` (the dictionaries are the count dictionaries), when you convert them into sorted lists you get: `sortedlist(a) = [0, 0, 1]` and `sortedlist(b) = [0, 1, 1]`. This loop + if statements are equivalent to:
```
for a_entry, b_entry in zip(sortedlist(a), sortedlist(b)):
    if a_entry != b_entry:
        return a_entry < b_entry
```
i.e. it's like alphabetical ordering on the lists. The code in `__lt__` does this without actually converting these to lists by utilizing the multiset counts.

Ultimately we could have reversed this, but I naturally chose this ordering as it's how I typically generate lists of multisets by hand, e.g.:
```
Multiset([0, 0]), Multiset([0, 1]), Multiset([1, 1])
```
i.e. by incrementing entries analogous to increasing numbers in a place-value number system.

Reply:
Ah I see. I was specifically wondering about `>` vs `<`, but you have explained that you indeed want `>` because it indicates that lhs has more smaller things, ie is smaller.